api/osf_groups/serializers.py

```diff
@@ -0,0 +1,191 @@
+        return absolute_reverse(
+            'groups:group-member-detail',
+            kwargs={
+                'user_id': obj._id,
+                'group_id': self.context['request'].parser_context['kwargs']['group_id'],
+                'version': self.context['request'].parser_context['kwargs']['version'],
+            },
+        )
+
+
+class GroupMemberCreateSerializer(GroupMemberSerializer):
+    id = GroupCompoundIDField(source='_id', required=False, allow_null=True)
+    type = TypeField()
+    full_name = ser.CharField(required=False)
+    email = ser.EmailField(required=False, write_only=True)
+
+    def to_representation(self, instance, envelope='data'):
+        """
+        Use GroupMemberSerializer for the response, but GroupMemberCreateSerializer
+        for the request.  We only want full_name to be writable on create member (for unregistered members).
+        User serializer endpoints should be used to edit user's full_name.
+        """
+        return GroupMemberSerializer(instance=instance, context=self.context).data
+
+    def get_user_object(self, user_id, group):
+        if user_id:
+            user = OSFUser.load(user_id)
+            if not user:
+                raise exceptions.NotFound(detail='User with id {} not found.'.format(user_id))
+            if group.has_permission(user, 'member'):
+                raise exceptions.ValidationError(detail='User is already a member of this group.')
+            return user
+        return user_id
+
+    def create(self, validated_data):
+        group = self.context['group']
+        user = self.get_user_object(validated_data.get('_id', None), group)
+        auth = Auth(self.context['request'].user)
+        full_name = validated_data.get('full_name', None)
+        email = validated_data.get('email', None)
+        role = self.get_group_role(validated_data, MEMBER)
+
+        try:
+            if user:
+                self.get_member_method(group, role)(user, auth)
+            else:
+                if not full_name or not email:
+                    raise exceptions.ValidationError(detail='You must provide a full_name/email combination to add an unconfirmed member.')
+                else:
+                    user = group.add_unregistered_member(full_name, email, auth, role)
+        except ValueError as e:
+            # Making someone a member has the potential to violate the one-manager rule
+            raise exceptions.ValidationError(detail=str(e.message))
+
+        return user
+
+    class Meta:
+        type_ = 'group_members'
+
+
+class GroupMemberDetailSerializer(GroupMemberSerializer):
+    id = GroupCompoundIDField(source='_id', required=True)
+
+    def update(self, user, validated_data):
+        group = self.context['group']
+        role = self.get_group_role(validated_data, user.group_role(group))
+        auth = Auth(self.context['request'].user)
+
+        try:
+            # Making sure the one-manager rule isn't violated
```

Comment:
I understand in this context you don't want to down grade a manager by making them a member, thus removing all managers. 👍